frontends/p4/def_use.h

```diff
@@ -29,9 +29,24 @@ limitations under the License.
 
 namespace P4 {
 
+class ComputeWriteSet;
 class StorageFactory;
 class LocationSet;
 
+// A location in the program. Includes the context from the visitor, which needs to
+// be copied out of the Visitor::Context objects, as they are allocated on the stack and
+// will become invalid as the IR traversal continues
+struct loc_t {
+    const IR::Node *node;
+    const loc_t *parent;
+    bool operator==(const loc_t &a) const {
+        if (node != a.node) return false;
+        if (parent && a.parent) return *parent == *a.parent;
+        return parent == a.parent;
```

Comment:
You could short-circuit by putting `if (parent == a.parent) return true;` before the recursive parent call.

Reply:
Done